src/Tests/library/test_rwlocks.pl

```diff
@@ -0,0 +1,112 @@
+    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+
+:- asserta(user:file_search_path(library, './library')).
+:- use_module(library(rwlocks)).
+
+:- begin_tests(rwlocks).                                                                                                                                          
+                                                                                                                                                                      
+    test(priority_handling) :-
+        retractall(started(_)),
+        with_rwlock(test_lock, 
+                    (assertz(started(write)), 
+                    sleep(1), 
+                    writeln('Write operation completed')), 
+                    write),
+        with_rwlock(test_lock, 
+                    (assertz(started(read)), 
+                    sleep(2), 
+                    writeln('Read operation completed')), 
+                    read),
+        findall(X, started(X), Started),
+        writeln(Started),
+        assertion(Started == [write, read]).                                                                                                                    
+                                                                                                                                                                      
+    test(timeout_handling, [throws(time_limit_exceeded(rwlock))]) :-
+        thread_create(with_rwlock(test_lock, sleep(2), write), WriteThread, []),
+        sleep(0.1),  % Give some time for the write lock to be acquired
+        catch(with_rwlock(test_lock, true, write, [timeout(0.5)]),
+            time_limit_exceeded(rwlock),
+            ( thread_signal(WriteThread, throw(abort_timeout_test)),
+                retract_rwlock_data(test_lock),
+                rethrow(time_limit_exceeded(rwlock))
+            )),
+        thread_join(WriteThread, _).
+
+    % Helper predicate to clean up any leftover rwlock data
+    retract_rwlock_data(LockId) :-
+        retractall(access(LockId, _, _)),
+        retractall(waiting(LockId, _, _, _)).
+
+    % Helper predicate to re-throw the exception
+    rethrow(Exception) :-
+        throw(Exception).                                                                                                          
+                                                                                                                                                                      
+    test(concurrent_reads) :-
+        flag(concurrent_read_counter, _, 0),
```

Comment:
flag/3 is a bit old school and the name is global.    In this case I'd simply assert teh facts that the work is done.   Here you cannot use thread_local/1.   More in general, what tells me the tests ran concurrently?   That is not so easy.    You could test so by making them talk to each other instead.    Use a timeout to avoid the possibility they do not run concurrently and thus the thing deadlocks.